File: sort/quicksort/recipe-01/cxx/testquicksort2.cpp
#include <vector>
#include "random.hpp"
#include "sort_helper.hpp"

template <typename Container>
void swap(Container& lyst, int i, int j) {
    auto temp = lyst[i];
    lyst[i] = lyst[j];
    lyst[j] = temp;
}

template <typename Container>
int partition(Container& lyst, int left, int right) {
    int middle = (left + right) / 2;
    auto pivot = lyst[middle];
    lyst[middle] = lyst[right];
    lyst[right] = pivot;

    int boundary = left;
    for (int index = left; index < right; index++) {
        if (lyst[index] < pivot) {
            swap(lyst, index, boundary);
            boundary += 1;
        }
    }
    swap(lyst, right, boundary);
    return boundary;
}

template <typename Container>
void insertionSort(Container& lyst, int left, int right) {
    int i = left + 1;
    while (i <= right) {
        auto itemToInsert = lyst[i];
        int j = i - 1;
        while (j >= 0) {
            if (itemToInsert < lyst[j]) {
                lyst[j + 1] = lyst[j];
                j -= 1;
            } else {
                break;
            }
        }
        lyst[j + 1] = itemToInsert;
        i += 1;
    }
}

template <typename Container>
void quicksortHelper(Container& lyst, int left, int right) {
    if (right - left <= 50) {
        insertionSort(lyst, left, right);
    } else {
        int pivotLocation = partition(lyst, left, right);
        quicksortHelper(lyst, left, pivotLocation - 1);
        quicksortHelper(lyst, pivotLocation + 1, right);
    }
}

template <typename Container>
void quicksort(Container& lyst) {
    quicksortHelper(lyst, 0, lyst.size()-1);
}

int main() {
    int size = 1000;
    std::vector<int> lyst;
    Random random;
    for (int count = 0; count < size; count++) {
        lyst.push_back(random.randint(1, size + 1));
    }
    print(lyst);
    quicksort(lyst);
    print(lyst);
}

File: sort/quicksort/recipe-01/origin/testquicksort2.py
"""
Project 3.9
File: testquicksort.py

Tests the quicksort algorithm
"""

def insertionSort(lyst, left, right):
    i = left + 1
    while i <= right:
        itemToInsert = lyst[i]
        j = i - 1
        while j >= 0:
            if itemToInsert < lyst[j]:
                lyst[j + 1] = lyst[j]
                j -= 1
            else:
                break
        lyst[j + 1] = itemToInsert
        i += 1

def quicksort(lyst):
    quicksortHelper(lyst, 0, len(lyst) - 1)

def quicksortHelper(lyst, left, right):
    """Runs insertionSort if length of list
    segment is <= 50."""
    if right - left <= 50:
        insertionSort(lyst, left, right)
    else:
        pivotLocation = partition(lyst, left, right)
        quicksortHelper(lyst, left, pivotLocation - 1)
        quicksortHelper(lyst, pivotLocation + 1, right)

def partition(lyst, left, right):
    # Find the pivot and exchange it with the last item
    middle = (left + right) // 2
    pivot = lyst[middle]
    lyst[middle] = lyst[right]
    lyst[right] = pivot
    # Set boundary point to first position
    boundary = left
    # Move items less than pivot to the left
    for index in range(left, right):
        if lyst[index] < pivot:
            swap(lyst, index, boundary)
            boundary += 1
    # Exchange the pivot item and the boundary item
    swap (lyst, right, boundary)
    return boundary

def swap(lyst, i, j):
    """Exchanges the items at positions i and j."""
    # You could say lyst[i], lyst[j] = lyst[j], lyst[i]
    # but the following code shows what is really going on
    temp = lyst[i]
    lyst[i] = lyst[j]
    lyst[j] = temp


import random


def main(size = 1000, sort = quicksort):
    lyst = []
    for count in range(size):
        lyst.append(random.randint(1, size + 1))
    print(lyst)
    sort(lyst)
    print(lyst)

if __name__ == "__main__":
    main() 
